File: csharp/1020.number-of-enclaves.cs
/*
 * @lc app=leetcode id=1020 lang=csharp
 *
 * [1020] Number of Enclaves
 */

using System.Linq;

// @lc code=start
public class Solution {
    private static void Walk(int[][] A, int m, int n, int i, int j) {
        if (i < 0 || i >= m || j < 0 || j >= n || A[i][j] == 0) {
            return;
        }
        A[i][j] = 0;
        Walk(A, m, n, i, j - 1);
        Walk(A, m, n, i, j + 1);
        Walk(A, m, n, i - 1, j);
        Walk(A, m, n, i + 1, j);
    }
    
    public int NumEnclaves(int[][] A) {
        int m = A.Length, n = A[0].Length;
        for (int i = 0; i < m; ++i) {
            Walk(A, m, n, i, 0);
            Walk(A, m, n, i, n - 1);
        }
        for (int j = 0; j < n; ++j) {
            Walk(A, m, n, 0, j);
            Walk(A, m, n, m - 1, j);
        }
        return A
            .SelectMany(r => r)
            .Count(x => x == 1);
    }
}
// @lc code=end

File: java/1020.number-of-enclaves.java
/*
 * @lc app=leetcode id=1020 lang=java
 *
 * [1020] Number of Enclaves
 */

// @lc code=start
class Solution {
    public int numEnclaves(int[][] grid) {
        int m = grid.length, n = grid[0].length, ret = 0;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; ++i) {
            visit(grid, i, 0, visited);
            visit(grid, i, n - 1, visited);
        }
        for (int i = 0; i < n; ++i) {
            visit(grid, 0, i, visited);
            visit(grid, m - 1, i, visited);
        }
        for (int i = 1; i < m - 1; ++i) {
            for (int j = 1; j < n - 1; ++j) {
                if (grid[i][j] == 1 && !visited[i][j]) {
                    ret += visit(grid, i, j, visited);
                }
            }
        }
        return ret;
    }

    private static int visit(int[][] grid, int i, int j, boolean[][] visited) {
        if (i < 0 || i >= visited.length || j < 0 || j >= visited[0].length) {
            return 0;
        }
        if (grid[i][j] == 0) {
            return 0;
        }
        if (visited[i][j]) {
            return 0;
        }
        visited[i][j] = true;
        return 1 + visit(grid, i - 1, j, visited) + visit(grid, i + 1, j, visited) + visit(grid, i, j - 1, visited) + visit(grid, i, j + 1, visited);
    }
}
// @lc code=end

